modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -2013,94 +1589,67 @@ public CompletableFuture<TableImpl> tableImplAsync(String name) {
+        CompletionListener<Void> tablesListener = (token, v, th) -> {
+            if (th == null) {
+                CompletableFuture<Map<Integer, TableImpl>> tablesFuture = tablesByIdVv.get(token);
 
-                    tablesFut.whenComplete((tables, e) -> {
-                        if (e != null) {
-                            getTblFut.completeExceptionally(e);
-                        } else {
-                            TableImpl table = tables.get(id);
+                tablesFuture.whenComplete((tables, e) -> {
+                    if (e != null) {
+                        getLatestTableFuture.completeExceptionally(e);
+                    } else {
+                        TableImpl table = tables.get(tableId);
 
-                            if (table != null) {
-                                getTblFut.complete(table);
-                            }
+                        if (table != null) {
+                            getLatestTableFuture.complete(table);
                         }
-                    });
-                } else {
-                    getTblFut.completeExceptionally(th);
-                }
-            };
-
-            assignmentsUpdatedVv.whenComplete(tablesListener);
+                    }
+                });
+            } else {
+                getLatestTableFuture.completeExceptionally(th);
+            }
+        };
 
-            // This check is needed for the case when we have registered tablesListener,
-            // but tablesByIdVv has already been completed, so listener would be triggered only for the next versioned value update.
-            tbl = latestTablesById().get(id);
+        assignmentsUpdatedVv.whenComplete(tablesListener);
 
-            if (tbl != null) {
-                assignmentsUpdatedVv.removeWhenComplete(tablesListener);
+        // This check is needed for the case when we have registered tablesListener,
+        // but tablesByIdVv has already been completed, so listener would be triggered only for the next versioned value update.
+        tableImpl = latestTablesById().get(tableId);
 
-                return completedFuture(tbl);
-            }
+        if (tableImpl != null) {
+            assignmentsUpdatedVv.removeWhenComplete(tablesListener);
 
-            return getTblFut.whenComplete((unused, throwable) -> assignmentsUpdatedVv.removeWhenComplete(tablesListener));
-        }));
-    }
+            return completedFuture(tableImpl);
+        }
 
-    /**
-     * Checks that the table is configured with specific id.
-     *
-     * @param id Table id.
-     * @return True when the table is configured into cluster, false otherwise.
-     */
-    private boolean isTableConfigured(int id) {
-        return configuredTablesCache.isTableConfigured(id);
+        return anyOf(getLatestTableFuture, stopManagerFuture)
+                .thenComposeAsync(o -> getLatestTableFuture, ioExecutor)
```

Comment:
what's the point of using `thenComposeAsync` instead of `thenCompose` here? Also, what's the point of `anyOf` above, which uses the same future, if we still need to wait for it?

Reply:
> what's the point of using thenComposeAsync instead of thenCompose here?

I thought that it would be better when one of the futures is completed (from `anyOf`), the subsequent processing of the futures continued to be performed in the inner pool, so as not to block the threads that complete the futures, but maybe I'm really bothering, wdyt?

> Also, what's the point of anyOf above, which uses the same future, if we still need to wait for it?

I will try to explain: if we `busyLock#enterBusy` somewhere and do not release it for some time, then in order to avoid blocking on the `TableManager#stop`, we will simply immediately complete the future with an error.
We discussed the decision with @sanpwc and decided to keep it.

Reply:
> I thought that it would be better when one of the futures is completed (from anyOf), the subsequent processing of the futures continued to be performed in the inner pool

This is ok, but I think it's a bit redundant, because looks like there's not a lot of things to execute, we wait for a future and then remove a listener